File: backend/orchestrator/storage.py
import json
import threading
from pathlib import Path
from datetime import datetime
import hashlib

lock = threading.Lock()

AGENTS_FILE = Path("agents.json")
EXPERIMENTS_FILE = Path("experiments.json")
USER_AGENT_MAPPING_FILE = Path("user_agent_mapping.json")

for file in [AGENTS_FILE, EXPERIMENTS_FILE]:
    if not file.exists():
        file.write_text(json.dumps({}))


def read_json(file_path):
    with lock:
        return json.loads(file_path.read_text())


def write_json(file_path, data):
    with lock:
        file_path.write_text(json.dumps(data, indent=4))

def user_agent_mapping(agent_id, user_id, verification_token):
    user_agent_mapping = read_json(USER_AGENT_MAPPING_FILE)
    user_agent_mapping[user_id] = {
        "agent_id": agent_id,
        "verification_token": hashlib.sha256(verification_token.encode()).hexdigest()
    }
    write_json(USER_AGENT_MAPPING_FILE, user_agent_mapping)

def verify_token(verification_token):
    user_agent_mapping = read_json(USER_AGENT_MAPPING_FILE)
    for user in user_agent_mapping.keys():
        if user_agent_mapping[user]["verification_token"] == hashlib.sha256(verification_token.encode()).hexdigest():
            return user, user_agent_mapping[user]["agent_id"]
    return "Invalid verification token", "Invalid verification token"


def register_agent(agent_id, host):
    agents = read_json(AGENTS_FILE)
    agents[agent_id] = {
        "host": host,
        "last_seen": datetime.utcnow().isoformat()
    }
    write_json(AGENTS_FILE, agents)


def update_agent_last_seen(agent_id):
    agents = read_json(AGENTS_FILE)
    if agent_id in agents:
        agents[agent_id]["last_seen"] = datetime.utcnow().isoformat()
        write_json(AGENTS_FILE, agents)

def create_experiment(exp_id, data):
    experiments = read_json(EXPERIMENTS_FILE)
    experiments[exp_id] = {
        **data,
        "status": "pending",
        "assigned_to": data["agent_id"]
    }
    write_json(EXPERIMENTS_FILE, experiments)


def get_experiment_for_agent(agent_id):
    experiments = read_json(EXPERIMENTS_FILE)

    for exp_id, exp in experiments.items():
        if exp["status"] == "pending" and exp["assigned_to"] == agent_id:
            exp["status"] = "assigned"
            write_json(EXPERIMENTS_FILE, experiments)
            return {"experiment_id": exp_id, **exp}

    return None



def update_experiment_status(exp_id, status, result=None):
    experiments = read_json(EXPERIMENTS_FILE)

    if exp_id in experiments:
        experiments[exp_id]["status"] = status

        if result:
            experiments[exp_id]["result"] = result

        write_json(EXPERIMENTS_FILE, experiments)


def get_all_agents():
    return read_json(AGENTS_FILE)


def get_all_experiments():
    return read_json(EXPERIMENTS_FILE)

File: orchestrator/storage.go
package main

import (
	"encoding/json"
	"os"
	"time"
)

/* =========================
   FILE PATHS
========================= */

var mappingFile = "mapping.json"
var agentsFile = "agents.json"
var experimentsFile = "experiments.json"
var usersFile = "users.json"

/* =========================
   GENERIC JSON HELPERS
========================= */

func readJSON(file string) map[string]interface{} {

	data := make(map[string]interface{})

	f, err := os.ReadFile(file)
	if err != nil {
		return data
	}

	json.Unmarshal(f, &data)
	return data
}

func writeJSON(file string, data map[string]interface{}) {

	bytes, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(file, bytes, 0644)
}

/* =========================
   USER → AGENT MAPPING
========================= */

func storeUserAgentMapping(userID, agentID, token string) {

	data := readJSON(mappingFile)

	if _, ok := data[userID]; !ok {
		data[userID] = []interface{}{}
	}

	list := data[userID].([]interface{})

	list = append(list, map[string]interface{}{
		"agent_id":           agentID,
		"verification_token": token,
	})

	data[userID] = list

	writeJSON(mappingFile, data)
}

/* =========================
   TOKEN VERIFICATION
========================= */

func verifyToken(token string) (string, string) {

	data := readJSON(mappingFile)

	for userID, v := range data {

		list := v.([]interface{})

		for _, item := range list {

			m := item.(map[string]interface{})

			if m["verification_token"] == token {
				return userID, m["agent_id"].(string)
			}
		}
	}

	return "", ""
}

/* =========================
   AGENT STORAGE
========================= */

func registerAgentStore(agentID, host string) {

	data := readJSON(agentsFile)

	data[agentID] = map[string]interface{}{
		"host":      host,
		"last_seen": time.Now().String(),
		"status":    "active",
	}

	writeJSON(agentsFile, data)
}

func updateLastSeen(agentID string) {

	data := readJSON(agentsFile)

	if agent, ok := data[agentID]; ok {

		m := agent.(map[string]interface{})
		m["last_seen"] = time.Now().String()
		m["status"] = "active"

		data[agentID] = m
		writeJSON(agentsFile, data)
	}
}

/* =========================
   EXPERIMENT STORAGE
========================= */

func storeExperiment(id string, exp interface{}) {

	data := readJSON(experimentsFile)

	expMap := map[string]interface{}{}

	bytes, _ := json.Marshal(exp)
	json.Unmarshal(bytes, &expMap)

	expMap["status"] = "pending"
	expMap["created_at"] = time.Now().String()

	data[id] = expMap

	writeJSON(experimentsFile, data)
}

func getExperimentForAgent(agentID string) map[string]interface{} {

	data := readJSON(experimentsFile)

	for id, v := range data {

		exp := v.(map[string]interface{})

		if exp["agent_id"] == agentID && exp["status"] == "pending" {

			exp["status"] = "running"
			data[id] = exp
			writeJSON(experimentsFile, data)

			exp["experiment_id"] = id
			return exp
		}
	}

	return nil
}

func updateExperimentStatus(id, status string, result map[string]interface{}) {

	data := readJSON(experimentsFile)

	if exp, ok := data[id]; ok {

		m := exp.(map[string]interface{})
		m["status"] = status
		m["result"] = result
		m["completed_at"] = time.Now().String()

		data[id] = m
		writeJSON(experimentsFile, data)
	}
}
